uefi/src/proto/network/snp.rs

```diff
@@ -0,0 +1,587 @@
+//! Simple Network Protocol
+//!
+//! Provides a packet level interface to a network adapter.
+//! Once the adapter is initialized, the protocol provides services that allows
+//! packets to be transmitted and received.
+//!
+//! No interface function must be called until `SimpleNetwork.start` is successfully
+//! called first.
+
+use super::{IpAddress, MacAddress};
+use crate::data_types::Event;
+use crate::{Result, Status};
+use core::ffi::c_void;
+use core::ptr;
+use core::ptr::NonNull;
+use uefi_macros::{unsafe_guid, Protocol};
+
+/// The Simple Network Protocol
+#[repr(C)]
+#[unsafe_guid("a19832b9-ac25-11d3-9a2d-0090273fc14d")]
+#[derive(Protocol)]
+pub struct SimpleNetwork {
+    revision: u64,
+    start: extern "efiapi" fn(this: &Self) -> Status,
+    stop: extern "efiapi" fn(this: &Self) -> Status,
+    initialize: extern "efiapi" fn(
+        this: &Self,
+        extra_recv_buffer_size: usize,
+        extra_transmit_buffer_size: usize,
+    ) -> Status,
+    reset: extern "efiapi" fn(this: &Self, extended_verification: bool) -> Status,
+    shutdown: extern "efiapi" fn(this: &Self) -> Status,
+    receive_filters: extern "efiapi" fn(
+        this: &Self,
+        enable: u32,
+        disable: u32,
+        reset_mcast_filter: bool,
+        mcast_filter_count: usize,
+        mcast_filter: Option<NonNull<MacAddress>>,
+    ) -> Status,
+    station_address:
+        extern "efiapi" fn(this: &Self, reset: bool, new: Option<&MacAddress>) -> Status,
+    statistics: extern "efiapi" fn(
+        this: &Self,
+        reset: bool,
+        stats_size: Option<&mut usize>,
+        stats_table: Option<&mut NetworkStats>,
+    ) -> Status,
+    mcast_ip_to_mac:
+        extern "efiapi" fn(this: &Self, ipv6: bool, ip: &IpAddress, mac: &mut MacAddress) -> Status,
+    nv_data: extern "efiapi" fn(
+        this: &Self,
+        read_write: bool,
+        offset: usize,
+        buffer_size: usize,
+        buffer: *mut c_void,
+    ) -> Status,
+    get_status: extern "efiapi" fn(
+        this: &Self,
+        interrupt_status: Option<&mut InterruptStatus>,
+        tx_buf: Option<&mut *mut c_void>,
+    ) -> Status,
+    transmit: extern "efiapi" fn(
+        this: &Self,
+        header_size: usize,
+        buffer_size: usize,
+        buffer: *const c_void,
+        src_addr: Option<&MacAddress>,
+        dest_addr: Option<&MacAddress>,
+        protocol: Option<&u16>,
+    ) -> Status,
+    receive: extern "efiapi" fn(
+        this: &Self,
+        header_size: Option<&mut usize>,
+        buffer_size: &mut usize,
+        buffer: *mut c_void,
+        src_addr: Option<&mut MacAddress>,
+        dest_addr: Option<&mut MacAddress>,
+        protocol: Option<&mut u16>,
+    ) -> Status,
+    // On QEMU, this event seems to never fire.
+    wait_for_packet: Event,
+    mode: *const NetworkMode,
+}
+
+impl SimpleNetwork {
+    /// Changes the state of a network from "Stopped" to "Started"
```

Comment:
General request here and elsewhere: comments and docstrings [should usually be full sentences](https://doc.rust-lang.org/beta/style-guide/index.html?highlight=sentence#comments), so ending in a period.

Also these docstrings should usually be written imperatively, i.e. "Change the state of a network" instead of "Changes the state of a network".